`unused_lifetimes` yeets unsafe binders

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->
Probably the most amusing bug I will encounter today :) 

I tried this code:

```rust
#![feature(unsafe_binders)]
#![allow(incomplete_features)]
#![warn(unused_lifetimes)]

use std::unsafe_binder::unwrap_binder;

#[derive(Copy, Clone)]
pub struct S([usize; 8]);

// Regression test for <https://github.com/rust-lang/rust/issues/141418>.
pub fn by_value(x: unsafe<'a> S) -> usize {
    unsafe { (|| unwrap_binder!(x).0[0])() }
}

fn main() {}

```

I expected to see this happen: *explanation*

Instead, this happened: *explanation*

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.89.0-nightly (1bbd62e54 2025-05-29)
binary: rustc
commit-hash: 1bbd62e547ba5cc08ccb44c27def3d33195d2dd5
commit-date: 2025-05-29
host: x86_64-unknown-linux-gnu
release: 1.89.0-nightly
LLVM version: 20.1.5
```
cargo check:
```
warning: lifetime parameter `'a` never used
  --> src/main.rs:11:27
   |
11 | pub fn by_value(x: unsafe<'a> S) -> usize {
   |                    -------^^--- help: elide the unused lifetime
   |
note: the lint level is defined here
  --> src/main.rs:3:9
   |
3  | #![warn(unused_lifetimes)]
   |         ^^^^^^^^^^^^^^^^
```

cargo fix
```
The following errors were reported:
error: expected type, found `)`
  --> src/main.rs:11:20
   |
11 | pub fn by_value(x: ) -> usize {
   |                    ^ expected type

error[E0425]: cannot find value `x` in this scope
  --> src/main.rs:12:33
   |
8  | pub struct S([usize; 8]);
   | ------------------------- similarly named tuple struct `S` defined here
...
12 |     unsafe { (|| unwrap_binder!(x).0[0])() }
   |                                 ^ help: a tuple struct with a similar name exists: `S`

error: aborting due to 2 previous errors
```

it removes the entire `unsafe<'a> S`  😆

--- a/compiler/rustc_resolve/src/late.rs
+++ b/compiler/rustc_resolve/src/late.rs
@@ -934,8 +934,7 @@ impl<'ra: 'ast, 'ast, 'tcx> Visitor<'ast> for LateResolutionVisitor<'_, 'ast, 'r
                 )
             }
             TyKind::UnsafeBinder(unsafe_binder) => {
-                // FIXME(unsafe_binder): Better span
-                let span = ty.span;
+                let span = ty.span.shrink_to_lo().to(unsafe_binder.inner_ty.span.shrink_to_lo());
                 self.with_generic_param_rib(
                     &unsafe_binder.generic_params,
                     RibKind::Normal,

--- a/tests/ui/unsafe-binders/unused-lifetimes-2.fixed
+++ b/tests/ui/unsafe-binders/unused-lifetimes-2.fixed
@@ -1,0 +1,20 @@
+// regression test for #141758
+//@ run-rustfix
+//@ check-pass
+
+#![warn(unused_lifetimes)]
+#![allow(incomplete_features, unused_imports, dead_code)]
+#![feature(unsafe_binders)]
+
+use std::unsafe_binder::unwrap_binder;
+
+#[derive(Copy, Clone)]
+pub struct S([usize; 8]);
+
+// Regression test for <https://github.com/rust-lang/rust/issues/141418>.
+pub fn by_value(_x: unsafe<'a> &'a S) -> usize {
+    //~^ WARN lifetime parameter `'b` never used
+    0
+}
+
+fn main() {}

--- a/tests/ui/unsafe-binders/unused-lifetimes-2.rs
+++ b/tests/ui/unsafe-binders/unused-lifetimes-2.rs
@@ -1,0 +1,20 @@
+// regression test for #141758
+//@ run-rustfix
+//@ check-pass
+
+#![warn(unused_lifetimes)]
+#![allow(incomplete_features, unused_imports, dead_code)]
+#![feature(unsafe_binders)]
+
+use std::unsafe_binder::unwrap_binder;
+
+#[derive(Copy, Clone)]
+pub struct S([usize; 8]);
+
+// Regression test for <https://github.com/rust-lang/rust/issues/141418>.
+pub fn by_value(_x: unsafe<'a, 'b> &'a S) -> usize {
+    //~^ WARN lifetime parameter `'b` never used
+    0
+}
+
+fn main() {}

--- a/tests/ui/unsafe-binders/unused-lifetimes-2.stderr
+++ b/tests/ui/unsafe-binders/unused-lifetimes-2.stderr
@@ -1,0 +1,16 @@
+warning: lifetime parameter `'b` never used
+  --> $DIR/unused-lifetimes-2.rs:15:32
+   |
+LL | pub fn by_value(_x: unsafe<'a, 'b> &'a S) -> usize {
+   |                              --^^
+   |                              |
+   |                              help: elide the unused lifetime
+   |
+note: the lint level is defined here
+  --> $DIR/unused-lifetimes-2.rs:5:9
+   |
+LL | #![warn(unused_lifetimes)]
+   |         ^^^^^^^^^^^^^^^^
+
+warning: 1 warning emitted
+

--- a/tests/ui/unsafe-binders/unused-lifetimes.fixed
+++ b/tests/ui/unsafe-binders/unused-lifetimes.fixed
@@ -1,0 +1,20 @@
+// regression test for #141758
+//@ run-rustfix
+//@ check-pass
+
+#![warn(unused_lifetimes)]
+#![allow(incomplete_features, unused_imports, dead_code)]
+#![feature(unsafe_binders)]
+
+use std::unsafe_binder::unwrap_binder;
+
+#[derive(Copy, Clone)]
+pub struct S([usize; 8]);
+
+// Regression test for <https://github.com/rust-lang/rust/issues/141418>.
+pub fn by_value(_x: S) -> usize {
+    //~^ WARN lifetime parameter `'a` never used
+    0
+}
+
+fn main() {}

--- a/tests/ui/unsafe-binders/unused-lifetimes.rs
+++ b/tests/ui/unsafe-binders/unused-lifetimes.rs
@@ -1,0 +1,20 @@
+// regression test for #141758
+//@ run-rustfix
+//@ check-pass
+
+#![warn(unused_lifetimes)]
+#![allow(incomplete_features, unused_imports, dead_code)]
+#![feature(unsafe_binders)]
+
+use std::unsafe_binder::unwrap_binder;
+
+#[derive(Copy, Clone)]
+pub struct S([usize; 8]);
+
+// Regression test for <https://github.com/rust-lang/rust/issues/141418>.
+pub fn by_value(_x: unsafe<'a> S) -> usize {
+    //~^ WARN lifetime parameter `'a` never used
+    0
+}
+
+fn main() {}

--- a/tests/ui/unsafe-binders/unused-lifetimes.stderr
+++ b/tests/ui/unsafe-binders/unused-lifetimes.stderr
@@ -1,0 +1,14 @@
+warning: lifetime parameter `'a` never used
+  --> $DIR/unused-lifetimes.rs:15:28
+   |
+LL | pub fn by_value(_x: unsafe<'a> S) -> usize {
+   |                     -------^^-- help: elide the unused lifetime
+   |
+note: the lint level is defined here
+  --> $DIR/unused-lifetimes.rs:5:9
+   |
+LL | #![warn(unused_lifetimes)]
+   |         ^^^^^^^^^^^^^^^^
+
+warning: 1 warning emitted
+

